packages/eslint-plugin/tests/rules/thenable-in-promise-aggregators.test.ts

```diff
@@ -0,0 +1,358 @@
+const messageIdArrayArg = 'arrayArg';
+const messageIdNonArrayArg = 'nonArrayArg';
+
+const ruleTester = new RuleTester({
+  parserOptions: {
+    ecmaVersion: 2018,
+    tsconfigRootDir: rootDir,
+    project: './tsconfig.json',
+  },
+  parser: '@typescript-eslint/parser',
+});
+
+ruleTester.run('thenable-in-promise-aggregators', rule, {
+  valid: [
+    `
+async function test() {
+  await Promise.race([Promise.resolve(3)]);
+}
+    `,
+    `
+async function test() {
+  await Promise.all([Promise.resolve(3)]);
+}
+    `,
+    `
+async function test() {
+  await Promise.allSettled([Promise.resolve(3)]);
+}
+    `,
+    `
+async function test() {
+  await Promise.race([(async () => true)()]);
+}
+    `,
+    `
+async function test() {
+  function returnsPromise() {
+    return Promise.resolve('value');
+  }
+  await Promise.race([returnsPromise()]);
+}
+    `,
+    `
+async function test() {
+  async function returnsPromiseAsync() {}
+  await Promise.race([returnsPromiseAsync()]);
+}
+    `,
+    `
+async function test() {
+  let anyValue: any;
+  await Promise.race([anyValue]);
+}
+    `,
+    `
+async function test() {
+  let unknownValue: unknown;
+  await Promise.race([unknownValue]);
+}
+    `,
+    `
+async function test() {
+  const numberPromise: Promise<number>;
+  await Promise.race([numberPromise]);
+}
+    `,
+    `
+async function test() {
+  class Foo extends Promise<number> {}
+  const foo: Foo = Foo.resolve(2);
+  await Promise.race([foo]);
+
+  class Bar extends Foo {}
+  const bar: Bar = Bar.resolve(2);
+  await Promise.race([bar]);
+}
+    `,
+    `
+async function test() {
+  await Promise.race([Math.random() > 0.5 ? numberPromise : 0]);
+  await Promise.race([Math.random() > 0.5 ? foo : 0]);
+  await Promise.race([Math.random() > 0.5 ? bar : 0]);
+
+  const intersectionPromise: Promise<number> & number;
+  await Promise.race([intersectionPromise]);
+}
+    `,
+    `
+async function test() {
+  class Thenable {
+    then(callback: () => {}) {}
+  }
+  const thenable = new Thenable();
+
+  await Promise.race([thenable]);
+}
+    `,
+    `
+const doSomething = async (
+  obj1: { a?: { b?: { c?: () => Promise<void> } } },
+  obj2: { a?: { b?: { c: () => Promise<void> } } },
+  obj3: { a?: { b: { c?: () => Promise<void> } } },
+  obj4: { a: { b: { c?: () => Promise<void> } } },
+  obj5: { a?: () => { b?: { c?: () => Promise<void> } } },
+  obj6?: { a: { b: { c?: () => Promise<void> } } },
+  callback?: () => Promise<void>,
+): Promise<void> => {
+  await Promise.all([
+    obj1.a?.b?.c?.(),
+    obj2.a?.b?.c(),
+    obj3.a?.b.c?.(),
+    obj4.a.b.c?.(),
+    obj5.a?.().b?.c?.(),
+    obj6?.a.b.c?.(),
+  ]);
+
+  await Promise.allSettled([callback?.()]);
+};
+    `,
+    `
+async function test() {
+  const promiseArr: Promise<number>[];
+  await Promise.all(promiseArr);
+}
+    `,
+    `
+async function test() {
+  const intersectionArr: (Promise<number> & number)[];
+  await Promise.all(intersectionArr);
+}
+    `,
```

Comment:
Maybe we should also add some cases with unions? Because it's not clear to me whether this rule should report on

```ts
declare const something: Promise<number> | number
Promise.all([something])
```

Reply:
Adding explicit tests for unions. Personally I think it makes sense to let union types through, though I could see a case for a flag for it. I think unioning a promise with a non-promise is somewhat rare, but when it does happen it's intentional. For example, I do this regularly on buttons wrappers in react:

```tsx
type MyButtonProps = {
  /**
   * Handler for when the button is clicked. If the result is a promise, the button will be disabled
   * while the promise is running. If the promise takes a long time, the screen will be darkened slightly
   * and an animation will play
   */
  onClick: () => Promise<void> | undefined;
}
```

Then it's a question of if the caller should check for undefined first, or just run it through the promise aggregator and use the fact it'll resolve instantly. Here I'd be fine being explicit, but I wonder if there are some more complicated cases where that'd be tedious

Reply:
Yeah I think unions should be allowed. It's a nice convenience and intentionally part of the Promise aggregator design to allow this.